// @ybd-project/ytdl-core/package/core/clients/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TvEmbedded = exports.Tv = exports.Ios = exports.Android = exports.MWeb = exports.WebEmbedded = exports.WebCreator = exports.Web = void 0;
var Web_1 = require("./Web");
Object.defineProperty(exports, "Web", { enumerable: true, get: function () { return __importDefault(Web_1).default; } });
var WebCreator_1 = require("./WebCreator");
Object.defineProperty(exports, "WebCreator", { enumerable: true, get: function () { return __importDefault(WebCreator_1).default; } });
var WebEmbedded_1 = require("./WebEmbedded");
Object.defineProperty(exports, "WebEmbedded", { enumerable: true, get: function () { return __importDefault(WebEmbedded_1).default; } });
var MWeb_1 = require("./MWeb");
Object.defineProperty(exports, "MWeb", { enumerable: true, get: function () { return __importDefault(MWeb_1).default; } });
var Android_1 = require("./Android");
Object.defineProperty(exports, "Android", { enumerable: true, get: function () { return __importDefault(Android_1).default; } });
var Ios_1 = require("./Ios");
Object.defineProperty(exports, "Ios", { enumerable: true, get: function () { return __importDefault(Ios_1).default; } });
var Tv_1 = require("./Tv");
Object.defineProperty(exports, "Tv", { enumerable: true, get: function () { return __importDefault(Tv_1).default; } });
var TvEmbedded_1 = require("./TvEmbedded");
Object.defineProperty(exports, "TvEmbedded", { enumerable: true, get: function () { return __importDefault(TvEmbedded_1).default; } });
